source/common/stats/stats_impl.cc

```diff
@@ -122,6 +122,71 @@ RawStatData* HeapRawStatDataAllocator::alloc(const std::string& name) {
   return data;
 }
 
+TagProducerImpl::TagProducerImpl(const envoy::api::v2::StatsConfig& config) : TagProducerImpl() {
+  // To check name conflict.
+  std::unordered_set<std::string> names;
+  reserveResources(config);
+  addDefaultExtractors(config, names);
+
+  for (const auto& tag_specifier : config.stats_tags()) {
+    if (!names.emplace(tag_specifier.tag_name()).second) {
+      throw EnvoyException(fmt::format("Tag name '{}' specified twice.", tag_specifier.tag_name()));
+    }
+
+    // If no tag value are found, fallback to default regex to keep backward compatibility.
+    if (tag_specifier.tag_value_case() == envoy::api::v2::TagSpecifier::TAG_VALUE_NOT_SET ||
+        tag_specifier.tag_value_case() == envoy::api::v2::TagSpecifier::kRegex) {
+      tag_extractors_->emplace_back(Stats::TagExtractorImpl::createTagExtractor(
+          tag_specifier.tag_name(), tag_specifier.regex()));
+
+    } else if (tag_specifier.tag_value_case() == envoy::api::v2::TagSpecifier::kFixedValue) {
+      default_tags_->emplace_back(
+          Stats::Tag{tag_specifier.tag_name(), tag_specifier.fixed_value()});
+    }
+  }
+}
+
+TagProducerImpl::TagProducerImpl()
+    : tag_extractors_(ExtractorsPtr{new std::vector<TagExtractorPtr>()}),
+      default_tags_(TagsPtr{new std::vector<Tag>()}) {}
+
+TagProducerImpl::~TagProducerImpl() {}
+
+std::string TagProducerImpl::produceTags(const std::string& name, std::vector<Tag>& tags) const {
+  if (!default_tags_->empty()) {
```

Comment:
nit: do we need this `if()`?  It should be fine to insert an empty vector, right?

Reply:
Right. My concern was the cost of inserting an emptry vector.